File: Search2DMatrix/search_2D_matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        for row in matrix:
            if target <= row[-1]:
                break
        lb, ub = 0, len(row) - 1

        while lb <= ub:
            guess = lb + (ub - lb)//2
            if row[guess] < target:
                lb = guess + 1
            elif target < row[guess]:
                ub = guess - 1
            else:
                return True
        return False
            
            

File: Search2DMatrix/search_2D_matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int l = 0, r = matrix.size()*matrix[0].size()-1;
        while (l <= r){
            int mid = l + (r - l)/2;
            int row = mid/matrix[0].size(), col = mid%matrix[0].size();
            if (matrix[row][col] < target){
                l = mid + 1;
            }
            else if (target < matrix[row][col]){
                r = mid - 1;
            }
            else{
                return true;
            }
        }   
        return false;
    }
};

